python/mpt/src/mpt/trie_diff.py

```diff
@@ -0,0 +1,388 @@
+                pre_code_hash = Hash32.fromhex(diff["preAccount"]["codeHash"][2:])
+                pre_storage_hash = Hash32.fromhex(diff["preAccount"]["storageHash"][2:])
+                pre_account = AccountNode(
+                    nonce=pre_nonce,
+                    balance=pre_balance,
+                    code_hash=pre_code_hash,
+                    storage_root=pre_storage_hash,
+                )
+            else:
+                pre_account = None
+
+            if "postAccount" in diff:
+                post_balance = U256(int(diff["postAccount"]["balance"][2:], 16))
+                post_nonce = Uint(int(diff["postAccount"]["nonce"][2:], 16))
+                post_code_hash = Hash32.fromhex(diff["postAccount"]["codeHash"][2:])
+                post_storage_hash = Hash32.fromhex(
+                    diff["postAccount"]["storageHash"][2:]
+                )
+                post_account = AccountNode(
+                    nonce=post_nonce,
+                    balance=post_balance,
+                    code_hash=post_code_hash,
+                    storage_root=post_storage_hash,
+                )
+            else:
+                post_account = None
+
+            if "storage" in diff:
+                for storage_diff in diff["storage"]:
+                    key = Bytes32.fromhex(storage_diff["storageKey"][2:])
+                    pre_int = int(storage_diff["preValue"][2:], 16)
+                    post_int = int(storage_diff["postValue"][2:], 16)
+                    pre = None if pre_int == 0 else U256(pre_int)
+                    post = None if post_int == 0 else U256(post_int)
+                    if address not in state_diff._storage_tries:
+                        state_diff._storage_tries[address] = {}
+                    state_diff._storage_tries[address][key] = tuple((pre, post))
+
+            state_diff._main_trie[address] = tuple((pre_account, post_account))
+
+        return state_diff
+
+    @classmethod
+    def from_tries(cls, tries: EthereumTrieTransitionDB) -> "StateDiff":
+        # Merge all mappings of hash -> node to have a single DB to fetch nodes and resolve addresses from
+        diff = StateDiff(
+            {}, {}, tries.nodes, tries.address_preimages, tries.storage_key_preimages
+        )
+
+        l_root = tries.state_root
+        r_root = tries.post_state_root
+
+        diff._compute_diff(l_root, r_root, Bytes(), diff._process_account_diff)
+        return diff
+
+    def _compute_diff(
+        self,
+        left: Optional[Hash32 | Bytes],
+        right: Optional[Hash32 | Bytes],
+        path: Bytes,
+        process_leaf_diff: Callable,
+    ):
+        if left == right:
+            return
+
+        def resolve(node: Hash32 | Bytes | None | InternalNode) -> InternalNode | None:
+            if isinstance(node, InternalNode):
+                return node
+            if not node or node == b"":
+                return None
+            if len(node) == 32:
```

Comment:
what is the behaviour of calling len on an Internal Node?

Reply:
cannot happen because of condition 1